### PR TITLE
feature: Convert new H2B Derek Dataset

### LINKED ISSUE #12: Add documentation + support for background images
- Add description of backdrops to the `DATA_FORMAT` documentation.
- Add fields to the metadata writer to accept and write backdrop images (including key + name).
- Update data scripts to include background images (for EMT H2B project, using dataset provided by @derekthirstrup )

### PR DESCRIPTION
**Pull request by ShrimpCryptid**
Closes #12 and #22! This PR updates the H2B labelfree conversion script to handle Derek's new dataset, and makes a bunch of usability improvements to the writer utility package in the process.

*Expected review time: large, ~30-40 min*

New features:
- Adds `write_categorical_feature`, which can automatically detect categories from an array of string (or object) data.
- Adds two utilities for working with backdrops, `copy_and_add_backdrops` and `add_backdrops`. Copying is parallelized!
- The writer will load in any existing manifest if it exists and only update changed fields, preserving fields like `bounds`, `frames`, and `backdrops`. This solves #22.

### CODE REVIEW

File: colorizer_data/writer.py [hunk 1]
```diff
@@ -130,22 +140,76 @@ class ColorizerDatasetWriter:
...
-    def write_feature(self, data: np.ndarray, info: FeatureInfo):
+        # Check for existence of manifest file in expected path
+        manifest_path = self.outpath + "/manifest.json"
+        if os.path.exists(manifest_path) and not force_overwrite:
+            # Load in the existing file
+            try:
+                with open(manifest_path, "r") as f:
+                    self.manifest = json.load(f)
+                logging.info(
+                    "An existing manifest file was found in the output directory and will be updated."
```
Comment (toloudis) [inline colorizer_data/writer.py:165]: 👍

File: colorizer_data/writer.py [hunk 2]
```diff
@@ -130,22 +140,76 @@ class ColorizerDatasetWriter:
...
+        # Clear features
+        self.manifest["features"] = []
+
+        # Load backdrops from existing manifest, if applicable
+        self.backdrops = {}
+        if "backdrops" in self.manifest:
+            for backdrop_metadata in self.manifest["backdrops"]:
+                self.backdrops[backdrop_metadata["key"]] = backdrop_metadata
+
+    def write_categorical_feature(self, data: np.ndarray, info: FeatureInfo) -> None:
```
Comment (toloudis) [inline colorizer_data/writer.py:184]: why can't `write_feature` just call this internally?  Then client scripts will not have to write their own logic to distinguish and can still just call write_feature for everything

File: colorizer_data/bin/example_scripts/convert_emt_h2b_labelfree.py [hunk 3]
```diff
@@ -296,25 +378,30 @@ def make_features(
...
+            if units == "um":
+                units = "µm"
+            info = FeatureInfo(
+                label=label,
+                unit=units,
+                column_name=feature_column,
+                type=FeatureType.CONTINUOUS,
+            )
+            # Auto-detect categorical features
+            if data.dtype.kind in {"U", "S", "O"}:  # unicode, string, or object
```
Comment (toloudis) [inline colorizer_data/bin/example_scripts/convert_emt_h2b_labelfree.py:401]: seems like this logic should just be inside of write_feature

File: colorizer_data/utils.py [hunk 4]
```diff
@@ -239,3 +244,61 @@ def generate_frame_paths(
...
+            filename_count[filename] = count + 1
+        else:
+            relative_paths.append(os.path.join(subdir_path, filename))
+            filename_count[filename] = 0
+    return relative_paths
+
+
+def copy_remote_or_local_file(src_path: str, dst_path: str) -> None:
+    """Copies a source file from a URL or filepath to destination filename."""
+    if src_path.startswith("http"):
```
Comment (toloudis) [inline colorizer_data/utils.py:290]: this looks like a use case we don't have yet.  did you test with http files?

File: colorizer_data/bin/example_scripts/convert_emt_h2b_labelfree.py (hunk 3 above)
Comment (ShrimpCryptid) [reply colorizer_data/bin/example_scripts/convert_emt_h2b_labelfree.py:403]: Moved it in #30! The changes were sizeable enough (and needed unit tests) so they're now in their own PR. I'll merge them all in a stack when everything is approved.

File: colorizer_data/writer.py (hunk 2 above)
Comment (ShrimpCryptid) [reply colorizer_data/writer.py:184]: Resolved in #30!

**Review by toloudis — APPROVED** [review]
LGTM knowing that the write_feature changes are coming after

**Pull request merged**